internal/rsm/statemachine.go

```diff
@@ -341,12 +344,28 @@ func (s *StateMachine) recover(ss pb.Snapshot, init bool) error {
 		}
 		s.applyOnDisk(ss, init)
 	} else {
+		// A snapshot can only be applied if we have at least the OnDiskIndex of the snapshot reached.
+		// If we have not reached it, a recovery above would be needed.
+		if init {
+			if ss.OnDiskIndex > s.onDiskInitIndex {
+				plog.Panicf("%s, ss.OnDiskIndex (%d) > s.onDiskInitIndex (%d)",
+					s.id(), ss.OnDiskIndex, s.onDiskInitIndex)
+			}
+		} else if ss.OnDiskIndex > s.onDiskIndex {
```

Comment:
@lni I don't know if this assumption is right? I thought mit would be consistent with `RecoverRequired` What do you think?

Reply:
I think the logic is correct but it is expressing something like 

```
if A {
} else {
 if A {
  panic()
 }
}
```

I also found the existing code to be easier to read & understand.

When it checks "if ss.Witness || ss.Dummy", it checks to see whether it is a partial snapshot image that does not contains user State Machine data. When true, it chooses the fast path to only apply the snapshot metadata. Everything below that are used to handle situation in which you do have a full snapshot to apply. In your changed version, it is no longer that clearly separated and obvious. For example, for your line 349 and 354, I have to carefully double check what conditions can actually lead to there. 

I'd recommend to just add that check inside the if ss.Witness || ss.Dummy {} block without changing the struct of the remaining part of the StateMachine.recover() method.

For the comments in line 347-348, could you please explicitly mention that this check is actually for situations like data corruption. Thanks.

Reply:
I see your point regarding the fast path. Nevertheless I have two considerations if we would simply add the new check inside the `if ss.Witness || ss.Dummy {}` block.

- I find it somehow confusing that in this edge case and for a dummy snapshot `recoverRequired` (in its original version) would simply return `true`, although a dummy snapshot can never be recovered.
- What about a shrunk snapshot? In its original version `recoverRequired` would return `false` for a shrunk snapshot during init, which would go to the path to apply the metadata on it. Would it be correct to apply the metadata of a shrunk snapshot which would hurt our new invariant?

Reply:
Partial snapshots (including witness and dummy snapshots) are not suppose to reach recoverRequired as in the original implementation they would be handled by that ss.Witness || ss.Dummy {} block. 

That if ss.Witness || ss.Dummy check tests whether the snapshot is a partial one that doesn't have the snapshot content. We should probably include checking for shrunk snapshots as well - as that is another type of partial snapshot. Once confirmed to be a partial snapshot (witness, dummy or shunk snapshot), we can apply its metadata, check our new invariant and exit. 

We can probably run that check again in recoverRequired to make sure we don't accidentally call recoverRequired on partial snapshots in the future. The drawback is that checking whether a snapshot is a shrunk one involves disk IO - it is totally fine for my NVME boxes but might be a concern for some people.

For shrunk snapshot, they are similar to dummy snapshot, the major difference between them is that dummy has this dummy field set to true in its pb.Snapshot record on its creation. A shrunk one doesn't have that, disk IO is involved to check whether shrunk or not. A snapshot for an IOnDiskStateMachine only becomes a shrunk one after it has been fully restored into the state machine. It doesn't have the dummy field set to be true because there is no easy way to atomically swap both of its snapshot file on disk and its pb.Snapshot in LogDB with a shrunk version. Our new invariant still holds because when we have a shrunk snapshot, it implies that its content has been fully restored into the state machine, in that case, on restart (init=true), s.onDiskInitIndex < ss.OnDiskIndex implies corruption.

Reply:
I've updated the PR, what do you think? I'll fix/update tests as soon as you agree with the state of the PR.

Reply:
It looks good to me. One minor issue, please make that comment at line 339 no more than 80 chars per line. Thanks.

Reply:
Awesome, I've updated the PR with a fix for the comment line and also added some tests.